billing/models.py

```diff
@@ -1,3 +1,91 @@
+from billing.common.strings import BillingValidationErrorText
+from managers.models import Project
+from users.common.fields import ChoiceEnum
+from users.models import CustomUser
+from utils.models import BaseModel
+
+
+class BillingDetailsAbstractQuerySet(models.QuerySet):
+    pass
+
+
+class BillingDetailsAbstract(BaseModel):
+    class Meta:
+        abstract = True
+
+    class Currency(ChoiceEnum):
+        PLN = CurrencyConstants.CurrencyDict["PLN"]["Code"]
+        USD = CurrencyConstants.CurrencyDict["USD"]["Code"]
+        EUR = CurrencyConstants.CurrencyDict["EUR"]["Code"]
+
+    class AccountingMethod(ChoiceEnum):
+        HOURLY = AccountingMethodText.HOURLY
+        MONTHLY = AccountingMethodText.MONTHLY
+
+    user = models.ForeignKey(CustomUser, on_delete=models.PROTECT, related_name="%(class)s")
+    value = models.PositiveSmallIntegerField(BillingDetailsAbstractModelText.VALUE, blank=False, null=False)
+    currency = models.CharField(
+        max_length=BillingConstants.CURRENCY_MAX_LENGTH.value, choices=Currency.choices(), default=Currency.PLN
+    )
+    accounting_method = models.CharField(
+        max_length=BillingConstants.ACCOUNTING_METHOD_MAX_LENGTH.value,
+        choices=AccountingMethod.choices(),
+        default=AccountingMethod.HOURLY.name,
+    )
+    from_date = models.DateTimeField(
+        BillingDetailsAbstractModelText.FROM_DATE, auto_now_add=False, blank=False, null=False
+    )
+    to_date = models.DateTimeField(BillingDetailsAbstractModelText.TO_DATE, auto_now_add=False, blank=True, null=True)
+
+    objects = BillingDetailsAbstractQuerySet.as_manager()
+
+    def clean(self) -> None:
+        super().clean()
+
+        if self.to_date is not None and self.to_date <= self.from_date:
+            raise ValidationError(message=BillingValidationErrorText.VALIDATION_ERROR_FROM_DATE_LATER_THAN_TO_DATE)
+        if self.to_date is None and not self.user.is_active:
+            raise ValidationError(message=BillingValidationErrorText.VALIDATION_ERROR_TO_DATE_EMPTY_FOR_INACTIVE_USER)
+
+    @property
+    def get_currency_name(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Name"]
+
+    @property
+    def get_currency_code(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Code"]
+
+    @property
+    def get_currency_symbol(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Symbol"]
+
+
+class UserSalaryDetails(BillingDetailsAbstract):
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        self._fix_user_has_two_ongoing_salaries()
+        super(UserSalaryDetails, self).save(*args, **kwargs)
+
+    def _fix_user_has_two_ongoing_salaries(self) -> None:
+        if self.to_date is None:
+            if UserSalaryDetails.objects.filter(user_id=self.user.pk).filter(to_date__isnull=True).exists():
+                salary_details = UserSalaryDetails.objects.filter(user_id=self.user.pk).get(to_date__isnull=True)
```

Comment:
I see that you're using it multiple times. You can create from this Manager and use it without writing again and again this queries.

Please create `UserSalaryDetailsManager`. I'll show you example of that:

```python
class UserSalaryDetailsManager(Manager):
    def get_ongoing_salary_or_none(self):
        try:
            return self.get(user_id=self.user.id, to_date__isnull=True)
        except self.model.DoesNotExist:
            return None
```
and then you need to attach this manager to model:

```python
class UserSalaryDetails(BaseModel):
    objects = UserSalaryDetailsManager()
```

And by this you will be able to use it everywhere where you want.

In this case you will be able to use it that way:
```python
salary_details = UserSalaryDetails.objects.get_ongoing_salary_or_none()
```

Reply:
Done